### PR TITLE
Move setting of 'optimizer' GUC to the extension

### PR DESCRIPTION
**Pull request by whitehawk**
Move setting of 'optimizer' GUC to the extension

Problem:
File 'guc_gp.c' contained a compile-time dependency on Orca, related to
'optimizer' GUC. It prevented moving of all Orca's code into a shared lib and
making its work transparent to the GPDB core.

This patch:
1. Sets the default value for 'optimizer' GUC to 'false'.
2. Adds enabling of 'optimizer' GUC in gp_orca extension shared lib.
3. Updates check in 'check_optimizer()' function. Now it utilizes runtime check
of planner_hook presence to distinguish if an external planner is available or
not.

Note: the semantic of 'optimizer' GUC has changed. Previously it was used
solely to enable & disable Orca. Now it is used to enable & disable any general
external planner. The name of GUC hasn't been changed, because too many places
in code and infrastructure rely on it, and we aim not to blow up the size of
the patch.

### CODE REVIEW

**Comment by dkovalev1** [comment]
It looks good. The only question left is: will there be a possibility to have more than two optimizers? As we know, PG12 also has geqo at least. I understand that this may be premature, and not relevant for this particular task but anyway.

**Comment by whitehawk** [comment]
> It looks good. The only question left is: will there be a possibility to have more than two optimizers? As we know, PG12 also has geqo at least. I understand that this may be premature, and not relevant for this particular task but anyway.

We can support many planners, but not simultaneously in runtime. The decision which one to use now can be done by switching between shared libraries.  I think that now we do not aim to support Standard + 2 external planners at the same time.

**Review by dkovalev1 — APPROVED** [review]

**Review by andr-sokolov — DISMISSED** [review]

**Comment by whitehawk** [comment]
Fixed conflicts.

File: src/backend/utils/misc/guc_gp.c [hunk 1]
```diff
@@ -5158,13 +5155,17 @@ check_gp_resource_group_bypass(bool *newval, void **extra, GucSource source)
 static bool
 check_optimizer(bool *newval, void **extra, GucSource source)
 {
-#ifndef USE_ORCA
-	if (*newval)
+	/*
+	 * Assume that, if no planner_hook is registered, we can use only standard
+	 * Postgres planner. Thus, forbid setting the GUC.
+	 */
+	if ((GP_ROLE_DISPATCH == Gp_role) &&
```
Comment (RekGRpth) [inline src/backend/utils/misc/guc_gp.c]: So, it turns out that we are returning `true` on the executors?

Comment (whitehawk) [reply src/backend/utils/misc/guc_gp.c]: That's the behaviour we had before this patch.

Comment (RekGRpth) [reply src/backend/utils/misc/guc_gp.c]: without ORCA on executors this function returned false before this patch

**Review by andr-sokolov — APPROVED** [review]

**Review by dkovalev1 — APPROVED** [review]

**Pull request merged**